message/validation/validator.go

```diff
@@ -0,0 +1,467 @@
+			Signers: make(map[spectypes.OperatorID]*SignerState),
+		}
+	}
+	return mv.index[id]
+}
+
+func (mv *MessageValidator) validateSignerBehavior(
+	state *ConsensusState,
+	signer spectypes.OperatorID,
+	msg *queue.DecodedSSVMessage,
+) error {
+	signedMsg, ok := msg.Body.(*specqbft.SignedMessage)
+	if !ok {
+		// TODO: add support
+		return fmt.Errorf("not supported yet")
+	}
+
+	signerState := state.Signers[signer]
+	if signerState == nil {
+		signerState = &SignerState{}
+	}
+
+	// Validate slot.
+	messageSlot := phase0.Slot(signedMsg.Message.Height)
+	if signerState.Slot > messageSlot {
+		// Signers aren't allowed to decrease their slot.
+		// If they've sent a future message due to clock error,
+		// this should be caught by the earlyMessage check.
+		return errors.New("signer has already advanced to a later slot")
+	}
+
+	// Validate round.
+	msgRound := signedMsg.Message.Round
+	if signerState.Round > msgRound {
+		// Signers aren't allowed to decrease their round.
+		// If they've sent a future message due to clock error,
+		// they'd have to wait for the next slot/round to be accepted.
+		return errors.New("signer has already advanced to a later round")
+	}
+
+	if signerState.Slot < messageSlot && signerState.Round >= msgRound || signerState.Round < msgRound && signerState.Slot >= messageSlot {
+		return errors.New("if slot is in future, round must be also in future and vice versa")
+	}
+
+	role := msg.MsgID.GetRoleType()
+	maxRound := mv.maxRound(role)
+	if msgRound > maxRound {
+		return fmt.Errorf("round too high")
+	}
+
+	if msgRound-signerState.Round > allowedRoundsInFuture {
+		return fmt.Errorf("round is too far in the future, current %d, got %d", signerState.Round, msgRound)
+	}
+
+	estimatedRound := mv.currentEstimatedRound(role, phase0.Slot(signedMsg.Message.Height))
+	// msgRound - 2 <= estimatedRound <= msgRound + 1
+	if estimatedRound-msgRound > allowedRoundsInFuture || msgRound-estimatedRound > allowedRoundsInPast {
+		return fmt.Errorf("message round is too far from estimated, current %v, got %v", estimatedRound, msgRound)
+	}
+
+	if estimatedRound > maxRound {
+		return fmt.Errorf("estimated round too high")
+	}
+
+	// Advance slot & round, if needed.
+	if signerState.Slot < messageSlot || signerState.Round < msgRound {
+		signerState.Reset(messageSlot, msgRound)
+	}
+
+	// TODO: if this is a round change message, we should somehow validate that it's not being sent too frequently.
+
+	// TODO: move to MessageCounts?
+	if signerState.ProposalData == nil {
+		signerState.ProposalData = signedMsg.FullData
+	} else if !bytes.Equal(signerState.ProposalData, signedMsg.FullData) {
+		return fmt.Errorf("duplicated proposal with different data")
+	}
+
+	hashedFullData, err := specqbft.HashDataRoot(signedMsg.FullData)
+	if err != nil {
+		return fmt.Errorf("hash data root: %w", err)
+	}
+
+	if hashedFullData != signedMsg.Message.Root {
+		return fmt.Errorf("root doesn't match full data hash")
+	}
```

Comment:
Prepares and Commits don't use the FullData field. Is this checked just for proposal?